### ISSUE TITLE
Database Error during Geo-Query 

### ISSUE THREAD

**Issue by kzangeli** [Fiware/context.Orion-LD#953]
This issue was initiated by my dear colleagues in NEC, via e-mail:

From: Benjamin Hebgen <Benjamin.Hebgen@neclab.eu>
Sent: Montag, 6. Dezember 2021 14:54
To: Martin Bauer <Martin.Bauer@neclab.eu>
Subject: orion

Hi,

tested with (with and without @context)
```
{
  "id" : "urn:ngsi-ld:test123:99",
  "type" : "testtype123",
  "brandName" : {
    "type" : "Property",
    "address" : {
      "type" : "Property",
        "value" : {
          "city" : "Berlin",
          "street" : "Ulrich Strasse"
        }
     },
    "value" : [ 
      {
        "bla1" : "Mercedes"
      },
      {
        "bla2" : "Audi"
      }
   ],
    "unitCode" : "EUR"
  },
  "isParked" : {
    "type" : "Relationship",
    "providedBy" : {
      "type" : "Relationship",
      "object" : "urn:ngsi-ld:Person:Bob"
    },
    "object" :"urn:ngsi-ld:OffStreetParking:Downtown1",
    "observedAt" : "2017-07-29T12:00:04Z"
  },
  "location" : {
    "type" : "GeoProperty",
    "value" : {
      "type" : "Point",
      "coordinates" : [ 8.6447165, 49.4184066 ] 
    },
    "observedAt" :"2021-07-19T12:01:12.466Z"
  },
  "@context": "https://uri.etsi.org/ngsi-ld/v1/ngsi-ld-core-context.jsonld"
}

{
  "id": "urn:ngsi-ld:test:6",
  "type" : "testtype",
  "brandName": {
    "type" : "Property",
    "address" : {
      "type" : "Property",
      "value" : {
        "city" : "Berlin",
        "street" : "Ulrich Strasse"
      }
    },
    "value" : [
      {
       "bla1" : "Mercedes"
      },
      {
        "bla2" : "Audi"
      }
    ],
    "unitCode" : "EUR"
  },
  "isParked" : {
    "type" : "Relationship",
      "providedBy" : {
        "type" : "Relationship",
        "object" : "urn:ngsi-ld:Person:Bob"
      },
    "object" :"urn:ngsi-ld:OffStreetParking:Downtown1",
    "observedAt" : "2017-07-29T12:00:04Z"
  },
  "location" : {
    "type" : "GeoProperty",
    "value" : {
      "type" : "Polygon",
      "coordinates" : [ [
       [ 8.6447165, 49.4184066 ],
       [ 8.6447165, 49.4184066 ],
       [ 8.6447165, 49.4184066 ],
       [ 8.6447165, 49.4184066 ] 
      ] ]
    },
    "observedAt" :"2021-07-19T12:01:12.466Z"  
  },
  "@context": "https://uri.etsi.org/ngsi-ld/v1/ngsi-ld-core-context.jsonld"
}
```

2 errors thrown by orion-ld:

op=mongoCppLegacyGeoIndexCreate.cpp[60]:mongoCppLegacyGeoIndexCreate | msg=Database Error (error creating 2dsphere index for attribute 'location' for db 'orion')

Assertion for valid query (possibly an error message from the mongodb server):
```
"attr": {
  "error": {
    "code": 291,
    "codeName": "NoQueryExecutionPlans",
    "errmsg": "error processing query: ns=orion.entities ntoreturn=500\nTree: $and\n    _id.type $eq \"https://uri.etsi.org/ngsi-ld/default-context/testtype123\"\n    _id.id regex /.*/\n    GEONEAR  field=attrs.location.value maxdist=1500 isNearSphere=0\nSort: { creDate: 1 }\nProj: {}\n planner returned error :: caused by :: unable to find index for $geoNear query"},"namespace":"orion.entities","query":{"query":{"$or":[{"_id.id":{"$regularExpression":{"pattern":".*","options":""}},"_id.type":"https://uri.etsi.org/ngsi-ld/default-context/testtype123"}],"attrs.location.value":{"$near":{"$geometry":{"type":"Point","coordinates":[8.6447165,49.4184066]},"$maxDistance":1500.0}}},"orderby":{"creDate":1},"$readPreference":{"mode":"nearest"}}
  }
}

 

 

Tested geoquery:
```
 http://localhost:1026/ngsi-ld/v1/entities?type=testtype&geoproperty=location&georel=near;maxDistance==1500&geometry=Point&coordinates=
[8.6447165,49.4184066]&limit=500
```
 

MfG

Benni

**Comment by kzangeli** [comment]
I tested the requests but I got a different error. Might be cause we're with different versions of the mongodb server - I use v3.6.23 myself.
However, the error comes from the location polygon of entity urn:ngsi-ld:test:6, which as far as I understand is invalid.
It's just four identical points, while a polygon needs to (in my own words):
* have at least 4 points (fulfilled)
* start and end in the same point (also fulfilled)
* go in "counter-clockwise direction" (not fulfilled as 4 identical points)

I modified the polygon of urn:ngsi-ld:test:6 :: location and the creation of the entity worked just fine.

So, for this particular case, the broker should not accept the creation of the entity (with an invalid polygon) and it should instead respond with an error (400 Bad Request, ideally, but I have a feeling it might be a 500 Internal Error, as the mongo client lib is the one that detects the problem ...).

# So:
* how exactly did Orion-LD respond to these requests?
* and, if you fix the faulty polygon, does everything work?

**Comment by kzangeli** [comment]
Ah, also. The current implementation doesn't acknowledge the error at all. It just responds with a `201 Created`, but, no entity is actually created.  The bug has been fixed: #955

**Comment by martin-p-bauer** [comment]
Checking it again, we came to the same conclusion - the main problem was that ``201 Created`` was returned, but the entity was not created. If this bug has been fixed, as you say, the issue can be closed.

**Comment by kzangeli** [comment]
Fixed in PR #955
